### PR TITLE
Fix memory leak and fd leak

### PR DESCRIPTION
**Pull request by yubihong**
fd leak：
In explodeDDRPM(), we will break the while loop if fchmod() exec failed, which cause the open fd leak. So we close the fd before break.
memory leak：
The normal usage of dd_extract/dd_list is:
1.dd_extract [-vbmlfh] -k -d -r
2.dd_list [-vh] -k -d -a
But if we use dd_extract -k -k .../dd_list -k -k ...we will lost the first strdup string pionter which cause memory leak; And in the explodeDDRPM(), if we check the status of archive_open failed, we will retrun -1, which cause cpio leak;

### CODE REVIEW

**Comment by jkonecny12** [comment]
Hello, thanks for your contribution!
I have one question, don't we want to block parameter repeating in this case? From my understanding this is not expected behavior and raising an error instead seems to me like a better solution in this case to make it obvious that user shouldn't do this.

Is it useful to use these arguments multiple times?

**Comment by yubihong** [comment]
> Hello, thanks for your contribution!
> I have one question, don't we want to block parameter repeating in this case? From my understanding this is not expected behavior and raising an error instead seems to me like a better solution in this case to make it obvious that user shouldn't do this.
> 
> Is it useful to use these arguments multiple times?

Hello, @jkonecny12 ! Yes, your suggestion is another good solution. If one parameter repeat multiple times, in my patch, we can using the last parameter to continue the process. If you think raising an error is a better solution, I will push another patch.
And I have test using these arguments multiple times, it will enter the case multiple times.

**Comment by yubihong** [comment]
Hello, @jkonecny12, please review if this is what you wanted?

**Comment by jkonecny12** [comment]
I would rather prefer to raise an error. It seems to me that it is then easier to catch calling problem.

File: utils/dd/dd_extract.c [hunk 1]
```diff
@@ -155,14 +155,32 @@ int main(int argc, char *argv[])
             break;
 
         case 'd':
+            if (directory) {
+                logMessage(ERROR, "Multiple -d arguments\n");
+                show_help();
+                rc =1;
```
Comment (jkonecny12) [inline utils/dd/dd_extract.c]: Please fix the white spaces:
```
rc = 1;
```

**Comment by jkonecny12** [comment]
Jenkins, it's ok to test.

File: utils/dd/dd_extract.c (hunk 1 above)
Comment (yubihong) [reply utils/dd/dd_extract.c]: Done！

**Comment by yubihong** [comment]
> Looks great. I also tested this code and fixed our existing tests.
> 
> Thanks a lot for this contribution and pointing me on our broken tests!
You're welcome! I‘’m glad to contribute。

**Comment by yubihong** [comment]
> Looks great. I also tested this code and fixed our existing tests.
> 
> Thanks a lot for this contribution and pointing me on our broken tests!

You're welcome! I'm glad to contribute.

**Pull request merged**